### PR TITLE
Enable scanning more addresses

### PR DESCRIPTION
**Pull request by blurpesec**
[ch8102]
### Description
Enables scanning more addresses.

##### ToDo:
- [x] Refactor counter to disable adding more than 5 empty addresses.
- [x] Style better

### CODE REVIEW

File: src/components/LinkApp.tsx [hunk 1]
```diff
@@ -95,7 +95,7 @@ const LINK_VARIANTS = {
   }
 };
 
-const SLink = styled('a')<LinkStyleProps & React.AnchorHTMLAttributes<HTMLAnchorElement>>`
+export const SLink = styled('a')<LinkStyleProps & React.AnchorHTMLAttributes<HTMLAnchorElement>>`
```
Comment (FrederikBolding) [inline src/components/LinkApp.tsx:98]: There is no reason to export this, we should use `LinkApp`

File: src/services/WalletService/deterministic/useDeterministicWallet.ts [hunk 2]
```diff
@@ -92,10 +91,33 @@ const useDeterministicWallet = (
     }
   }, [state.isInit]);
 
+  // On first connection && on asset update
   useEffect(() => {
```
Comment (FrederikBolding) [inline src/services/WalletService/deterministic/useDeterministicWallet.ts:95]: Can we add tests for some of this? Goes for the entire feature. It is complex logic and one of the least tested parts of the app.

File: src/components/WalletUnlock/Hardware.tsx [hunk 3]
```diff
@@ -2,17 +2,8 @@ import React from 'react';
...
-  Button,
-  Heading,
-  Icon,
-  InlineMessage,
-  Spinner,
-  Text,
-  TIcon
-} from '@components';
+import { Box, BusyBottom, Button, Heading, InlineMessage, Spinner, Text, TIcon } from '@components';
+import Icon from '@components/Icon';
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/Hardware.tsx:6]: Import with the others

File: src/services/WalletService/deterministic/reducer.ts [hunk 4]
```diff
@@ -47,6 +45,7 @@ const DeterministicWalletReducer = (
   state: DeterministicWalletState,
   { type, payload, error }: DWAction
 ): DeterministicWalletState => {
+  console.debug('[', type, ']: triggered', state);
```
Comment (FrederikBolding) [inline src/services/WalletService/deterministic/reducer.ts:48]: Can you remove some of these? (multiple instances)

File: src/components/WalletUnlock/DeterministicWallet.tsx [hunk 5]
```diff
@@ -157,18 +185,8 @@ const DeterministicWallet = ({
...
-    const freshAddressGenerationSuccess = generateFreshAddress({
-      ...defaultDPath,
-      offset: freshAddressIndex,
-      numOfAddresses: 1
-    });
-    if (freshAddressGenerationSuccess) {
-      setFreshAddressIndex(freshAddressIndex + 1);
-    }
+  const handleScanMoreAddresses = (dpath: ExtendedDPath) => {
+    scanMoreAddresses!(dpath);
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicWallet.tsx:189]: Does this need to be possibly undefined?

File: src/components/WalletUnlock/DeterministicWallet.tsx [hunk 6]
```diff
@@ -247,35 +265,66 @@ const DeterministicWallet = ({
...
+            }}
+          />
+        </SBox>
+        <SBox
+          variant="columnAlignLeft"
+          style={{ visibility: !displayEmptyAddresses ? 'hidden' : 'visible' }}
+        >
+          <Text>
+            <Trans id="MNEMONIC_DPATH_SELECT" />{' '}
+            <LinkApp href="" onClick={() => setDpathAddView(true)}>
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicWallet.tsx:291]: I think `#` would be a valid non-problematic `href`

File: src/components/WalletUnlock/DeterministicWallet.tsx [hunk 7]
```diff
@@ -129,22 +146,33 @@ const initialFormikValues: FormValues = {
   value: ''
 };
 
+type TDPathOptionProps = OptionProps<DPath> | { data: DPath; selectOption?(): void };
+const DPathOption = ({ data, selectOption }: TDPathOptionProps) => (
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicWallet.tsx:150]: Can we move all the DPath selector code to a component?

File: src/components/WalletUnlock/DeterministicWallet.tsx [hunk 8]
```diff
@@ -247,35 +265,66 @@ const DeterministicWallet = ({
...
+            clearable={false}
+            searchable={false}
+          />
+        </SBox>
+      </Box>
+      <Box variant="rowAlign" justifyContent="space-between" pb={SPACING.SM}>
+        <Switch
+          checked={displayEmptyAddresses}
+          onChange={() => toggleDisplayEmptyAddresses(!displayEmptyAddresses)}
+          labelLeft="Hide Empty Addresses"
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicWallet.tsx:311]: Use translatable strings

File: src/components/WalletUnlock/DeterministicWallet.tsx [hunk 9]
```diff
@@ -247,35 +265,66 @@ const DeterministicWallet = ({
...
+            valueComponent={({ value }) => <DPathOption data={value} />}
+            clearable={false}
+            searchable={false}
+          />
+        </SBox>
+      </Box>
+      <Box variant="rowAlign" justifyContent="space-between" pb={SPACING.SM}>
+        <Switch
+          checked={displayEmptyAddresses}
+          onChange={() => toggleDisplayEmptyAddresses(!displayEmptyAddresses)}
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicWallet.tsx:310]: This isn't a toggle when you have to define the param. Maybe rename to `setDisplayEmptyAddresses`

File: src/components/WalletUnlock/DeterministicAccountList.tsx [hunk 10]
```diff
@@ -88,87 +80,120 @@ const SDownloader = styled(Downloader)`
...
   network,
-  freshAddressIndex,
-  generateFreshAddress,
+  displayEmptyAddresses,
+  selectedDPath,
+  handleScanMoreAddresses,
+  onUnlock,
   handleUpdate
 }: DeterministicAccountListProps) {
+  const MAX_EMPTY_ADDRESSES = 5;
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicAccountList.tsx:119]: Perhaps this should be defined more globally and imported?

File: src/components/WalletUnlock/DeterministicAccountList.tsx [hunk 11]
```diff
@@ -88,87 +80,120 @@ const SDownloader = styled(Downloader)`
...
 
-  const [selectedAccounts, setSelectedAccounts] = useState([] as ISelectedAccount[]);
-
-  const accountsToUse = uniqBy(prop('address'), finishedAccounts).filter(
-    ({ isFreshAddress, balance }) => (balance && !balance.isZero()) || isFreshAddress
-  );
+  const [tableAccounts, setTableAccounts] = useState({} as ITableAccounts);
 
+  const accountsToUse = uniqBy(prop('address'), finishedAccounts);
+  // setTableAccounts to be accountsToUse on update with isDefault set if it isn't already set and
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicAccountList.tsx:125]: And what? The suspense is killing me 😆

File: src/components/WalletUnlock/DeterministicAccountList.tsx [hunk 12]
```diff
@@ -88,87 +80,120 @@ const SDownloader = styled(Downloader)`
...
-    );
-    setSelectedAccounts(selected);
-  }, [accountsToUse.length]);
+    if (accountsToUse.length === 0 && Object.keys(tableAccounts).length !== 0) {
+      setTableAccounts({} as ITableAccounts);
+    } else {
+      const tableAccs = accountsToUse.reduce((acc, idx) => {
+        acc[idx.address] = tableAccounts[idx.address] || {
+          ...idx,
+          isDefaultConfig: true,
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicAccountList.tsx:133]: What does this param mean? Perhaps needs an explainer

File: src/components/WalletUnlock/DeterministicAccountList.tsx [hunk 13]
```diff
@@ -88,87 +80,120 @@ const SDownloader = styled(Downloader)`
...
+        ...tableAccounts,
+        [account.address]: {
+          ...account,
+          isDefaultConfig: false,
+          isSelected: !account.isSelected
+        }
+      });
+      return;
+    }
+    if (emptySelectedAccounts.length >= MAX_EMPTY_ADDRESSES) return;
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicAccountList.tsx:160]: Will a newly found account be added automatically since this assumes that you are only selecting empty accounts manually? Is this also the case if my account is not found initially and I have to keep scanning?

File: src/components/WalletUnlock/DeterministicAccountTable.tsx [hunk 14]
```diff
@@ -15,27 +15,32 @@ import {
...
 import { Downloader } from '../Downloader';
 
-interface DeterministicTableProps {
+export interface TableAccountDisplay extends DWAccountDisplay {
+  isSelected: boolean;
+  isDefaultConfig: boolean;
+}
+
+export interface ITableAccounts {
+  [key: string]: TableAccountDisplay;
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicAccountTable.tsx:32]: `Record<string, TableAccountDisplay>`

File: src/components/WalletUnlock/DeterministicAccountTable.tsx [hunk 15]
```diff
@@ -253,24 +258,45 @@ const SDownloader = styled(Downloader)`
   }
 `;
 
+const sortAccounts = (
```
Comment (FrederikBolding) [inline src/components/WalletUnlock/DeterministicAccountTable.tsx:261]: Move to helper & test

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: src/components/WalletUnlock/Hardware.tsx (hunk 3 above)
Comment (blurpesec) [reply src/components/WalletUnlock/Hardware.tsx:6]: Was doing this previously, but i think this was leading to a circular dependency in storybook builds

File: src/components/WalletUnlock/DeterministicAccountList.tsx (hunk 13 above)
Comment (blurpesec) [reply src/components/WalletUnlock/DeterministicAccountList.tsx]: Yep, left a comment there explaining this.

File: src/components/WalletUnlock/DeterministicAccountList.tsx (hunk 12 above)
Comment (blurpesec) [reply src/components/WalletUnlock/DeterministicAccountList.tsx]: removed

File: src/components/WalletUnlock/DeterministicAccountList.tsx (hunk 10 above)
Comment (blurpesec) [reply src/components/WalletUnlock/DeterministicAccountList.tsx]: ehhh its only used in this component, so i'm not sure what the point of moving it to our global configs would be.

**Comment by tayvano** [comment]
after scanning, changing derivation path doesnt seem to do anything and idk if that's normal
asset dropdown takes about a full second to respond and shows the blinking "|" like its ready for me to type in it, yet that field can't be typed in
perhaps add a grey check mark to all unselected accounts as an indicator that you can click there to select it
scanning and toggling empty/non-empty seems pretty good!
there's still a really, really terrible issue with windows 10 security popups that makes it so you cant interact with the screen for about a full minute during/after scanning, but idk if that's solvable. ledger has this article but it's completely useless because you can't just "move the security popup out of the way" because the security popup constantly opens and closes https://twitter.com/LSDinmycoffee/status/1371939585538818059

**Comment by tayvano** [comment]
Copied stuff from the massive slack thread below. not sure how much of this is still relevant. 

**I would strongly prefer to merge this sooner rather than later and create new cards for outstanding issues that need to be resolved.**

- after scanning, changing derivation path doesnt seem to do anything and idk if that's normal

- asset dropdown takes about a full second to respond and shows the blinking "|" like its ready for me to type in it, yet that field can't be typed in

- perhaps add a grey check mark to all unselected accounts as an indicator that you can click there to select it
scanning and toggling empty/non-empty seems pretty good!

- there's still a really, really terrible issue with windows 10 security popups that makes it so you cant interact with the screen for about a full minute during/after scanning, but idk if that's solvable. ledger has this article but it's completely useless because you can't just "move the security popup out of the way" because the security popup constantly opens and closes https://twitter.com/LSDinmycoffee/status/1371939585538818059

-after scanning, changing derivation path doesnt seem to do anything and idk if that's normal
I changed this to hide the derivation path dropdown unless empty addresses are shown (this is what is affected by the dpath dropdown).

- asset dropdown takes about a full second to respond and shows the blinking "|" like its ready for me to type in it, yet that field can't be typed in: I can't do much about this. The problem is that it takes a second to render 2000+ assets in a dropdown. All we can do is filter out some assets. but there's not a good way of determining which assets to allow users to scan by. Ideas ? perhaps add a grey check mark to all unselected accounts as an indicator that you can click there to select it meh. i don't think that's necessary atm since. we can add this in the future if it becomes an issue.

- Here ya go: https://app.bugreplay.com/#/shared/reports/b7a72934-687a-4407-8b8d-a07c57fc0b8d

- Yep. We don't even check whether WebUSB is supported for the wallet scanning for some reason.
https://github.com/MyCryptoHQ/MyCrypto/blob/master/src/services/WalletService/deterministic/DeterministicWalletService.tsx#L41-L45
But for the normal Ledger abstraction we check, which is used in signing and works there.
https://github.com/MyCryptoHQ/MyCrypto/blob/master/src/services/WalletService/deterministic/ledger.ts#L131-L138

- When scanning 10 more addresses is it only for a single DPath? If so, why is the DPath dropdown hidden until I look for empty addresses?

Jordan said the windows 10 issue seems to be resolved as of 5 minutes ago.

**Review by vilsbole — CHANGES_REQUESTED** [review]

File: src/services/WalletService/wallets/HardwareWallet.ts [hunk 16]
```diff
@@ -38,6 +39,14 @@ export default abstract class HWWallet implements Wallet {
 
   public abstract getDPaths(): DPath[];
 
+  /**
+   * Optional function that can be used to getMultipleAddresses from a device.
+   *
+   * @param {ExtendedDPath[]} paths The derivation paths to getMultipleAddresses.
+   * @return {Promise<any>} Can return an array of deterministic wallet's account display objects.
+   */
+  public abstract getMultipleAddresses(paths: ExtendedDPath[]): Promise<DWAccountDisplay[]>;
```
Comment (vilsbole) [inline src/services/WalletService/wallets/HardwareWallet.ts:48]: 😍

File: src/services/WalletService/wallets/ledger/Ledger.ts [hunk 17]
```diff
@@ -50,4 +51,38 @@ export default abstract class Ledger extends HardwareWallet {
...
+
+  public async getMultipleAddresses(dpaths: ExtendedDPath[]): Promise<DWAccountDisplay[]> {
+    if (dpaths.length === 0) {
+      console.error('[getMultipleAddresses]: Derivation paths not found');
+      return [];
+    }
+    return this.getDPathAddresses(dpaths);
+  }
+
+  public async getDPathAddresses(dpaths: ExtendedDPath[]): Promise<DWAccountDisplay[]> {
```
Comment (vilsbole) [inline src/services/WalletService/wallets/ledger/Ledger.ts:63]: 1. Is `getDpathAddresses` used elsewhere to justify it being separate from `getMultipleAddresses` ?

2. Test with `spyOn`
```
 let person = new Person('Lorem', 'Ipsum');
 let spy = jest.spyOn(person, 'sayMyName').mockImplementation(() => 'Hello');
```

3. Once tested, we can most likely optimise with `Promise.all`

File: src/services/WalletService/deterministic/useDeterministicWallet.ts [hunk 18]
```diff
@@ -27,117 +40,152 @@ const useDeterministicWallet = (
...
-          type: DWActionTypes.GET_ADDRESSES_SUCCESS
-        }),
-      handleEnqueueAccounts: (accounts: DWAccountDisplay[]) => {
-        return dispatch({
-          type: DWActionTypes.ENQUEUE_ADDRESSES,
-          payload: { accounts }
+  const getAccounts = async (session: Wallet, dpaths: ExtendedDPath[]) => {
+    // Trezor wallet uses getMultipleAddresses for fetching multiple addresses at a time. Ledger doesn't have this functionality.
+    if (session.getMultipleAddresses) {
+      await session
```
Comment (vilsbole) [inline src/services/WalletService/deterministic/useDeterministicWallet.ts:47]: FYI: No need to `await` when using `then`

File: src/services/WalletService/deterministic/useDeterministicWallet.ts [hunk 19]
```diff
@@ -27,117 +40,152 @@ const useDeterministicWallet = (
...
-        dispatch({
-          type: DWActionTypes.GET_ADDRESSES_SUCCESS
-        }),
-      handleEnqueueAccounts: (accounts: DWAccountDisplay[]) => {
-        return dispatch({
-          type: DWActionTypes.ENQUEUE_ADDRESSES,
-          payload: { accounts }
+  const getAccounts = async (session: Wallet, dpaths: ExtendedDPath[]) => {
+    // Trezor wallet uses getMultipleAddresses for fetching multiple addresses at a time. Ledger doesn't have this functionality.
+    if (session.getMultipleAddresses) {
```
Comment (vilsbole) [inline src/services/WalletService/deterministic/useDeterministicWallet.ts:46]: Prefer defensive form:
`if (!session.getMultipleAddress) throw new Error()`

It saves an indentation level.

File: src/components/WalletUnlock/components/DeterministicAccountTable.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+    value: '0.001'
+  },
+  csv: '',
+  onSelect: noOp,
+  handleUpdate: noOp,
+  handleScanMoreAddresses: noOp
+};
+
+export const DeterministicAccountTable = () => (
+  <ProvidersWrapper>
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/DeterministicAccountTable.stories.tsx:43]: Already provided in `preview.js`

File: src/components/WalletUnlock/components/DeterministicAccountTable.stories.tsx [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+import { fAssets, fDWAccounts, fNetworks } from '@fixtures';
+import { noOp } from '@utils';
+
+import {
+  default as DeterministicTable,
+  DeterministicTableProps,
+  ITableAccounts
+} from './DeterministicAccountTable';
+
+export default { title: 'Hardware/DeterministicAccountTable' };
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/DeterministicAccountTable.stories.tsx:15]: It's an `Organism`

File: src/components/WalletUnlock/components/DeterministicAccountTable.stories.tsx [hunk 22]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+
+import { ProvidersWrapper } from '@../jest_config/providersWrapper';
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/DeterministicAccountTable.stories.tsx:3]: import from `test-utils`

File: src/components/WalletUnlock/Hardware.tsx [hunk 23]
```diff
@@ -120,10 +109,11 @@ const HardwareWalletUI = ({ network, state, walletId, handleNullConnect }: Hardw
             />
           </ErrorMessageContainer>
         )}
+        <br />
         {state.isConnecting ? (
-          <div className="HardwarePanel-loading">
```
Comment (vilsbole) [inline src/components/WalletUnlock/Hardware.tsx:124]: 💌

File: src/services/WalletService/deterministic/useDeterministicWallet.ts [hunk 24]
```diff
@@ -27,117 +40,152 @@ const useDeterministicWallet = (
...
+    network: Network,
+    asset: ExtendedAsset
+  ) => {
+    const addresses = accounts.map(({ address }) => address);
+    const balanceLookup =
+      asset.type === 'base'
+        ? () => getBaseAssetBalancesForAddresses(addresses, network)
+        : () => getSingleTokenBalanceForAddresses(asset, network, addresses);
+
+    try {
```
Comment (vilsbole) [inline src/services/WalletService/deterministic/useDeterministicWallet.ts:77]: async patterns:
- when using `try/catch` prefer await.
- otherwise use `then/catch`.
Not both together.

File: src/services/WalletService/deterministic/reducer.ts [hunk 25]
```diff
@@ -47,6 +46,11 @@ const DeterministicWalletReducer = (
   state: DeterministicWalletState,
   { type, payload, error }: DWAction
 ): DeterministicWalletState => {
+  console.debug('[DeterministicWalletReducer]: state: ', state, '\naction: ', {
+    type,
+    payload,
+    error
+  });
```
Comment (vilsbole) [inline src/services/WalletService/deterministic/reducer.ts:53]: remove ?

**Review by vilsbole — COMMENTED** [review]
Nice! Now we just need to convert the reducer to a `slice` attach it to the `root.reducer` then convert the side-effects to sagas and add unit tests for each.

IMHO it's easier then trying to write a test for the hook.

File: src/config/dpaths.ts [hunk 26]
```diff
@@ -17,9 +17,8 @@ const ETH_LEDGER: DPath = {
 
 const ETH_LEDGER_LIVE: DPath = {
   label: 'Ledger Live (ETH)',
-  value: "m/44'/60'/0'/0/0",
-  isHardened: true,
-  getIndex: (addressIndex): string => `m/44'/60'/${addressIndex}'/0/0`
+  value: `m/44'/60'/addrIndex'/0/0`,
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply src/config/dpaths.ts:20]: We process derivation paths before using them (with `getFullPath`).

File: src/services/WalletService/deterministic/useDeterministicWallet.ts (hunk 2 above)
Comment (blurpesec) [reply src/services/WalletService/deterministic/useDeterministicWallet.ts]: dun did :3

File: src/components/WalletUnlock/components/DeterministicAccountTable.tsx [hunk 27]
```diff
@@ -371,37 +387,23 @@ const DeterministicTable = ({
               </LinkContainer>
             </Row>
           ))}
-          <GenerateAddressButton
-            onClick={generateFreshAddress}
-            disabled={!isComplete || freshAddressIndex >= DEFAULT_GAP_TO_SCAN_FOR}
+          <BottomActionButton
+            onClick={() => {
+              if (!isComplete) return;
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply src/components/WalletUnlock/components/DeterministicAccountTable.tsx]: it's a `span` not a button, since it's just clickable text so i don't think `disable` css works here

File: src/components/WalletUnlock/components/HDWList.tsx [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+
+import { Banner, Box, Button, Icon, Spinner, Tooltip, Typography } from '@components';
+import { DWAccountDisplay, ExtendedDPath } from '@services';
+import { BREAK_POINTS, COLORS, SPACING } from '@theme';
+import { Trans } from '@translations';
+import { BannerType, DPath, ExtendedAsset, Network } from '@types';
+import { accountsToCSV, bigify, useScreenSize } from '@utils';
+import { prop, uniqBy } from '@vendor';
+
+import { Downloader } from '../../Downloader';
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWList.tsx:13]: Can we import from `@components`?

File: src/components/WalletUnlock/components/HDWList.tsx [hunk 29]
```diff
@@ -0,0 +1,299 @@
...
+          ...account,
+          isSelected: !account.isSelected
+        }
+      });
+      return;
+    }
+    // disallows selecting an account that is empty if MAX_EMPTY_ADDRESSES is already met
+    if (
+      emptySelectedAccounts.length >= MAX_EMPTY_ADDRESSES &&
+      bigify(account.balance!).isEqualTo(0)
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWList.tsx:141]: move to utils fn: `hasBalance`?

File: src/components/WalletUnlock/components/HDWList.tsx [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+  color: ${COLORS.BLUE_MYC};
+  cursor: pointer;
+  font-weight: bold;
+  &:hover {
+    color: ${COLORS.BLUE_LIGHT_DARKISH};
+  }
+`;
+
+export const filterZeroBalanceAccounts = (accounts: DWAccountDisplay[]) =>
+  accounts.filter(({ balance }) => balance && bigify(balance).isZero());
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWList.tsx:74]: move to utils fn: `hasBalance`?

File: src/components/WalletUnlock/components/HDWList.tsx [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+  // setTableAccounts to be accountsToUse on update with isDefault set if it isn't already set
+  // and if accountsToUse is cleared (occurs when re-scanning all accounts or when changing asset), refresh tableAccounts
+  useEffect(() => {
+    if (accountsToUse.length === 0 && Object.keys(tableAccounts).length !== 0) {
+      setTableAccounts({} as ITableAccounts);
+    } else {
+      const tableAccs = accountsToUse.reduce((acc, idx) => {
+        acc[idx.address] = tableAccounts[idx.address] || {
+          ...idx,
+          isSelected: (idx.balance && !bigify(idx.balance).isZero()) || false
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWList.tsx:113]: move to utils fn: `hasBalance`?

File: src/components/WalletUnlock/components/HDWallet.tsx [hunk 32]
```diff
@@ -247,40 +236,64 @@ const DeterministicWallet = ({
...
-          }}
+
+      <Box pb={SPACING.MD} variant="spaceBetween">
+        <Box variant="columnAlignLeft" m={SPACING.XS}>
+          <Text>
+            <Trans id="MNEMONIC_SUBTITLE" />
+          </Text>
+          <AssetSelector
+            selectedAsset={assetToUse}
+            showAssetIcon={false}
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWallet.tsx:247]: Shouldn't we display the icon?

File: src/components/WalletUnlock/components/HDWallet.tsx [hunk 33]
```diff
@@ -157,21 +161,7 @@ const DeterministicWallet = ({
...
-      offset: freshAddressIndex,
-      numOfAddresses: 1
-    });
-    if (freshAddressGenerationSuccess) {
-      setFreshAddressIndex(freshAddressIndex + 1);
-    }
-  };
-
-  const csv = accountsToCSV(state.finishedAccounts, assetToUse);
+  const csv = accountsToCSV(scannedAccounts, assetToUse);
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWallet.tsx:164]: Move to selector.

File: src/components/WalletUnlock/components/HDWList.tsx [hunk 34]
```diff
@@ -0,0 +1,299 @@
...
+      return;
+    setTableAccounts({
+      ...tableAccounts,
+      [account.address]: {
+        ...account,
+        isSelected: !account.isSelected
+      }
+    });
+  };
+  const csv = accountsToCSV(scannedAccounts, asset);
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWList.tsx:152]: move to selector

File: src/services/Store/store/hdWallet.slice.ts [hunk 35]
```diff
@@ -0,0 +1,254 @@
...
+  const { asset, dpaths, network, walletId, setSession } = payload;
+  // initialize the wallet
+  try {
+    const session: Wallet = yield call(selectWallet, walletId);
+    yield call([session, 'initialize'], dpaths[0]);
+    yield put(slice.actions.requestConnection());
+    yield call(setSession, session);
+    yield put(slice.actions.requestConnectionSuccess({ asset, network }));
+  } catch (err) {
+    console.error(err);
```
Comment (vilsbole) [inline src/services/Store/store/hdWallet.slice.ts:199]: prefix error?

File: src/components/WalletUnlock/components/HDWList.tsx [hunk 36]
```diff
@@ -0,0 +1,299 @@
...
+  const { isMobile } = useScreenSize();
+
+  const [tableAccounts, setTableAccounts] = useState({} as ITableAccounts);
+
+  const accountsToUse = uniqBy(prop('address'), scannedAccounts);
+  // setTableAccounts to be accountsToUse on update with isDefault set if it isn't already set
+  // and if accountsToUse is cleared (occurs when re-scanning all accounts or when changing asset), refresh tableAccounts
+  useEffect(() => {
+    if (accountsToUse.length === 0 && Object.keys(tableAccounts).length !== 0) {
+      setTableAccounts({} as ITableAccounts);
```
Comment (vilsbole) [inline src/components/WalletUnlock/components/HDWList.tsx:108]: It would seem simpler to move this logic to the slice.

**Review by vilsbole — COMMENTED** [review]
Looking good!

For the UI:
1. We need to reset the state when the component dismounts.

Screenshot 2021-04-13 at 18 54 32
Screenshot 2021-04-13 at 18 56 04

File: src/components/WalletUnlock/components/HDWTable.tsx [hunk 37]
```diff
@@ -344,14 +361,14 @@ const DeterministicTable = ({
...
-                  <DPath>({account.pathItem.path})</DPath>
+                  <DPathDisplay>({account.pathItem.path})</DPathDisplay>
                 </DPathContainer>
                 <ValueContainer>
                   <Typography>
                     <Currency
                       amount={
                         account.balance
-                          ? bigify(fromTokenBase(account.balance, asset.decimal)).toFixed(4)
+                          ? bigify(fromTokenBase(bigify(account.balance), asset.decimal)).toFixed(4)
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply src/components/WalletUnlock/components/HDWTable.tsx:371]: `string`

File: src/components/WalletUnlock/components/HDWallet.tsx (hunk 32 above)
Comment (blurpesec) [reply src/components/WalletUnlock/components/HDWallet.tsx:247]: Just slows down the dropdown more. I'd say it'd be fine to add back the icon when we filter the assets list. But we don't have a way of separating important vs long-tail tokens, so it'd be better to avoid for now.

File: src/components/WalletUnlock/components/HDWList.tsx (hunk 28 above)
Comment (blurpesec) [reply src/components/WalletUnlock/components/HDWList.tsx:13]: I think this one results in cyclic imports errors

**Comment by tayvano** [comment]
not sure if me or bc i had multiple tabs open but i could successfully add my ledger accounts but not send with them

**Comment by jspence425** [comment]
> not sure if me or bc i had multiple tabs open but i could successfully add my ledger accounts but not send with them
> S 2021-04-22 at 12 45 40 PM

Adding in my anecdote: I just executed a TX with Ledger while using this PR in incognito. No problem here!

File: package.json [hunk 38]
```diff
@@ -241,7 +241,7 @@
     "prettier:diff": "prettier --write --config ./.prettierrc --list-different \"src/**/*.ts\" \"src/**/*.tsx\"",
     "update:tokens": "parse-eth-tokens --networks all --output ./src/database/data/tokens --exclude 0x5a276Aeb77bCfDAc8Ac6f31BBC7416AE1A85eEF2,0x0027449Bf0887ca3E431D263FFDeFb244D95b555",
     "lint:styles": "stylelint ./src/**/*.tsx",
-    "lint:code": "eslint . --ignore-path .gitignore --ext .ts,.tsx,.js,.jsx",
+    "lint:code": "eslint . --ignore-path .gitignore --ext .ts,.tsx,.js,.jsx --fix",
```
Comment (vilsbole) [inline package.json:244]: This is unrelated to PR. 
NB. It's recommended to configure your IDE to `fix` on save.

**Review by vilsbole — CHANGES_REQUESTED** [review]

**Comment by vilsbole** [comment]
Looking good!

Let's just:

1. Remove the spinner at the end of `...for the selected derivation path`.
Currently it's not aligned with the text. And since it has the same behaviour `Continuing to scan...` I suggest we remove it.
Screenshot 2021-04-27 at 16 21 56

2. Fix the index count in the banner: 
When I click on show hidden addresses, a blue banner is displayed mentioning that I can only add 5 accounts. Since the index count is wrong and the gating functionality incomplete I suggest we remove the banner and the associated code.

Screenshot 2021-04-27 at 16 22 53

**Review by vilsbole — APPROVED** [review]
